libdnf/rpm/rpm_signature.cpp

```diff
@@ -0,0 +1,210 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/rpm/rpm_signature.hpp"
+
+#include "repo/repo_gpgme.hpp"
+#include "rpm/rpm_log_guard.hpp"
+#include "utils/bgettext/bgettext-lib.h"
+#include "utils/bgettext/bgettext-mark-domain.h"
+#include "utils/fs/temp.hpp"
+#include "utils/url.hpp"
+
+#include "libdnf/repo/file_downloader.hpp"
+#include "libdnf/repo/repo.hpp"
+
+#include <rpm/rpmcli.h>
+#include <rpm/rpmdb.h>
+#include <rpm/rpmkeyring.h>
+#include <rpm/rpmlog.h>
+#include <rpm/rpmpgp.h>
+#include <rpm/rpmts.h>
+
+namespace libdnf::rpm {
+
+// deleters for unique_ptrs that take ownership of rpm objects
+// once the ownership is transfered, rpm objects get automatically freed when out of scope
+auto pkt_deleter = [](uint8_t * pkt) { free(pkt); };
+auto ts_deleter = [](rpmts_s * ts) { rpmtsFree(ts); };
+
+KeyInfo::KeyInfo(std::string key_url, const BaseWeakPtr & base) : key_url(key_url), base(base) {
+    std::unique_ptr<utils::fs::TempFile> downloaded_key;
+    if (utils::url::is_url(key_url)) {
+        if (key_url.starts_with("file://")) {
+            key_path = key_url.substr(7);
+        } else {
+            // download the remote key
+            downloaded_key = std::make_unique<libdnf::utils::fs::TempFile>("rpmkey");
+            libdnf::repo::FileDownloader downloader(base->get_config());
+            downloader.add(key_url, downloaded_key->get_path());
+            downloader.download(true, true);
+            key_path = downloaded_key->get_path();
+        }
+    } else {
+        key_path = key_url;
+    }
+
+    utils::fs::File key_file(key_path, "r");
+    for (auto & key_info : repo::RepoGpgme::rawkey2infos(key_file.get_fd())) {
+        key_id = key_info.get_id();
+        user_id = key_info.get_user_id();
+        fingerprint = key_info.get_fingerprint();
+    }
```

Comment:
What is the purpose of this loop?
`rawkey2infos` returns a vector with information about signing subkeys. The loop iterates through the vector and the next key information overwrites the previous one. So at the end we get information about the last signing subkey. Why is only the last subkey used? And why don't we  get the last subkey directly (the key from position size() - 1) instead of  looping through all the items in the vector?

Reply:
As I understand it, `rawkeys2infos` always extracts only the first signing subkey.
But I see the issue now - to reproduce it you need a file with multiple public keys. Having a key with multiple subkeys is not enough.
With such key file dnf4 will ask for confirmation of all the keys, current dnf5 implementation imports only the first one.
I'll fix it.